File: engine/src/renderer/vulkan_pipeline.c
#include "cardinal/core/log.h"
#include "vulkan_state.h"
#include <cardinal/renderer/vulkan_pipeline.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <vulkan/vulkan.h>

/**
 * @brief Creates depth resources for the Vulkan state.
 *
 * @param s Pointer to the VulkanState structure.
 * @return true if successful, false otherwise.
 *
 * @todo Support configurable depth formats and multisampling.
 * @todo Integrate with Vulkan dynamic rendering extensions.
 */
static bool create_depth_resources(VulkanState* s) {
    // Find a suitable depth format
    VkFormat candidates[] = {VK_FORMAT_D32_SFLOAT, VK_FORMAT_D32_SFLOAT_S8_UINT,
                             VK_FORMAT_D24_UNORM_S8_UINT};
    s->swapchain.depth_format = VK_FORMAT_UNDEFINED;

    for (int i = 0; i < 3; i++) {
        VkFormatProperties props;
        vkGetPhysicalDeviceFormatProperties(s->context.physical_device, candidates[i], &props);
        if (props.optimalTilingFeatures & VK_FORMAT_FEATURE_DEPTH_STENCIL_ATTACHMENT_BIT) {
            s->swapchain.depth_format = candidates[i];
            break;
        }
    }

    if (s->swapchain.depth_format == VK_FORMAT_UNDEFINED) {
        CARDINAL_LOG_ERROR("pipeline: failed to find suitable depth format");
        return false;
    }

    // Create depth image
    VkImageCreateInfo imageInfo = {0};
    imageInfo.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
    imageInfo.imageType = VK_IMAGE_TYPE_2D;
    imageInfo.extent.width = s->swapchain.extent.width;
    imageInfo.extent.height = s->swapchain.extent.height;
    imageInfo.extent.depth = 1;
    imageInfo.mipLevels = 1;
    imageInfo.arrayLayers = 1;
    imageInfo.format = s->swapchain.depth_format;
    imageInfo.tiling = VK_IMAGE_TILING_OPTIMAL;
    imageInfo.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;
    imageInfo.usage = VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT;
    imageInfo.samples = VK_SAMPLE_COUNT_1_BIT;
    imageInfo.sharingMode = VK_SHARING_MODE_EXCLUSIVE;

    // Use VulkanAllocator to allocate and bind image + memory
    if (!vk_allocator_allocate_image(&s->allocator, &imageInfo, &s->swapchain.depth_image,
                                     &s->swapchain.depth_image_memory,
                                     VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT)) {
        CARDINAL_LOG_ERROR("pipeline: allocator failed to create depth image");
        return false;
    }

    // Create depth image view
    VkImageViewCreateInfo viewInfo = {0};
    viewInfo.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
    viewInfo.image = s->swapchain.depth_image;
    viewInfo.viewType = VK_IMAGE_VIEW_TYPE_2D;
    viewInfo.format = s->swapchain.depth_format;
    viewInfo.subresourceRange.aspectMask = VK_IMAGE_ASPECT_DEPTH_BIT;
    viewInfo.subresourceRange.baseMipLevel = 0;
    viewInfo.subresourceRange.levelCount = 1;
    viewInfo.subresourceRange.baseArrayLayer = 0;
    viewInfo.subresourceRange.layerCount = 1;

    if (vkCreateImageView(s->context.device, &viewInfo, NULL, &s->swapchain.depth_image_view) !=
        VK_SUCCESS) {
        CARDINAL_LOG_ERROR("pipeline: failed to create depth image view");
        // Free image + memory via allocator on failure
        vk_allocator_free_image(&s->allocator, s->swapchain.depth_image,
                                s->swapchain.depth_image_memory);
        s->swapchain.depth_image = VK_NULL_HANDLE;
        s->swapchain.depth_image_memory = VK_NULL_HANDLE;
        return false;
    }

    // Initialize layout tracking
    s->swapchain.depth_layout_initialized = false;

    CARDINAL_LOG_INFO("pipeline: depth resources created");
    return true;
}

/**
 * @brief Destroys depth resources associated with the Vulkan state.
 *
 * @param s Pointer to the VulkanState structure.
 */
static void destroy_depth_resources(VulkanState* s) {
    if (!s || !s->context.device)
        return;

    // Validate and destroy depth image view
    if (s->swapchain.depth_image_view != VK_NULL_HANDLE) {
        vkDestroyImageView(s->context.device, s->swapchain.depth_image_view, NULL);
        s->swapchain.depth_image_view = VK_NULL_HANDLE;
    }

    // Validate and free image + memory using allocator
    if (s->swapchain.depth_image != VK_NULL_HANDLE ||
        s->swapchain.depth_image_memory != VK_NULL_HANDLE) {
        // Ensure allocator is valid before freeing
        if (s->allocator.device != VK_NULL_HANDLE) {
            vk_allocator_free_image(&s->allocator, s->swapchain.depth_image,
                                    s->swapchain.depth_image_memory);
        }
        s->swapchain.depth_image = VK_NULL_HANDLE;
        s->swapchain.depth_image_memory = VK_NULL_HANDLE;
    }

    // Reset layout tracking when depth resources are destroyed
    s->swapchain.depth_layout_initialized = false;
}

/**
 * @brief Creates the render pass and graphics pipeline for the Vulkan state.
 *
 * @param s Pointer to the VulkanState structure.
 * @return true if successful, false otherwise.
 *
 * @todo Support multiple render passes for advanced rendering techniques.
 * @todo Implement pipeline caching for faster recreation.
 */
bool vk_create_pipeline(VulkanState* s) {
    CARDINAL_LOG_INFO("pipeline: create depth resources");
    if (!create_depth_resources(s)) {
        return false;
    }

    CARDINAL_LOG_INFO("pipeline: depth resources created - no simple triangle pipeline needed");
    return true;
}

/**
 * @brief Destroys the render pass and graphics pipeline.
 *
 * @param s Pointer to the VulkanState structure.
 *
 * @todo Ensure thread-safe destruction of resources.
 * @todo Add logging for destruction events.
 */
void vk_destroy_pipeline(VulkanState* s) {
    if (!s || !s->context.device)
        return;

    // Wait for device to be idle before destroying resources for thread safety
    VkResult result = vkDeviceWaitIdle(s->context.device);
    if (result != VK_SUCCESS) {
        CARDINAL_LOG_ERROR("pipeline: vkDeviceWaitIdle failed during destruction: %d", result);
        // Continue with destruction anyway to prevent resource leaks
    }

    destroy_depth_resources(s);

    CARDINAL_LOG_INFO("pipeline: pipeline resources destroyed");
}

File: engine/src/renderer/vulkan_pipeline.zig
const std = @import("std");
const log = @import("../core/log.zig");
const types = @import("vulkan_types.zig");
const vk_allocator = @import("vulkan_allocator.zig");
const c = @import("vulkan_c.zig").c;
const memory = @import("../core/memory.zig");

const pipe_log = log.ScopedLogger("PIPELINE");

fn create_depth_resources(s: *types.VulkanState) bool {
    // Find a suitable depth format
    const candidates = [_]c.VkFormat{ c.VK_FORMAT_D32_SFLOAT, c.VK_FORMAT_D32_SFLOAT_S8_UINT, c.VK_FORMAT_D24_UNORM_S8_UINT };
    s.swapchain.depth_format = c.VK_FORMAT_UNDEFINED;

    for (candidates) |format| {
        var props: c.VkFormatProperties = undefined;
        c.vkGetPhysicalDeviceFormatProperties(s.context.physical_device, format, &props);
        if ((props.optimalTilingFeatures & c.VK_FORMAT_FEATURE_DEPTH_STENCIL_ATTACHMENT_BIT) != 0) {
            s.swapchain.depth_format = format;
            break;
        }
    }

    if (s.swapchain.depth_format == c.VK_FORMAT_UNDEFINED) {
        pipe_log.err("failed to find suitable depth format", .{});
        return false;
    }

    // Create depth image
    var imageInfo = std.mem.zeroes(c.VkImageCreateInfo);
    imageInfo.sType = c.VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
    imageInfo.imageType = c.VK_IMAGE_TYPE_2D;
    imageInfo.extent.width = s.swapchain.extent.width;
    imageInfo.extent.height = s.swapchain.extent.height;
    imageInfo.extent.depth = 1;
    imageInfo.mipLevels = 1;
    imageInfo.arrayLayers = 1;
    imageInfo.format = s.swapchain.depth_format;
    imageInfo.tiling = c.VK_IMAGE_TILING_OPTIMAL;
    imageInfo.initialLayout = c.VK_IMAGE_LAYOUT_UNDEFINED;
    imageInfo.usage = c.VK_IMAGE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT | c.VK_IMAGE_USAGE_SAMPLED_BIT;
    imageInfo.samples = c.VK_SAMPLE_COUNT_1_BIT;
    imageInfo.sharingMode = c.VK_SHARING_MODE_EXCLUSIVE;

    // Use VulkanAllocator to allocate and bind image + memory
    if (!vk_allocator.allocate_image(&s.allocator, &imageInfo, &s.swapchain.depth_image, &s.swapchain.depth_image_memory, &s.swapchain.depth_image_allocation, c.VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT)) {
        pipe_log.err("allocator failed to create depth image", .{});
        return false;
    }

    // Create depth image view
    var viewInfo = std.mem.zeroes(c.VkImageViewCreateInfo);
    viewInfo.sType = c.VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
    viewInfo.image = s.swapchain.depth_image;
    viewInfo.viewType = c.VK_IMAGE_VIEW_TYPE_2D;
    viewInfo.format = s.swapchain.depth_format;
    viewInfo.subresourceRange.aspectMask = c.VK_IMAGE_ASPECT_DEPTH_BIT;
    viewInfo.subresourceRange.baseMipLevel = 0;
    viewInfo.subresourceRange.levelCount = 1;
    viewInfo.subresourceRange.baseArrayLayer = 0;
    viewInfo.subresourceRange.layerCount = 1;

    if (c.vkCreateImageView(s.context.device, &viewInfo, null, &s.swapchain.depth_image_view) != c.VK_SUCCESS) {
        pipe_log.err("failed to create depth image view", .{});
        // Free image + memory via allocator on failure
        vk_allocator.free_image(&s.allocator, s.swapchain.depth_image, s.swapchain.depth_image_allocation);
        s.swapchain.depth_image = null;
        s.swapchain.depth_image_memory = null;
        return false;
    }

    // Initialize layout tracking
    s.swapchain.depth_layout_initialized = false;

    pipe_log.info("depth resources created", .{});
    return true;
}

fn destroy_depth_resources(s: *types.VulkanState) void {
    if (s.context.device == null) return;

    // Validate and destroy depth image view
    if (s.swapchain.depth_image_view != null) {
        c.vkDestroyImageView(s.context.device, s.swapchain.depth_image_view, null);
        s.swapchain.depth_image_view = null;
    }

    // Validate and free image + memory using allocator
    if (s.swapchain.depth_image != null or s.swapchain.depth_image_memory != null) {
        // Ensure allocator is valid before freeing
        if (s.allocator.device != null) {
            vk_allocator.free_image(&s.allocator, s.swapchain.depth_image, s.swapchain.depth_image_allocation);
        }
        s.swapchain.depth_image = null;
        s.swapchain.depth_image_memory = null;
    }

    // Reset layout tracking when depth resources are destroyed
    s.swapchain.depth_layout_initialized = false;
}

fn create_pipeline_cache(s: *types.VulkanState) bool {
    var cache_info = std.mem.zeroes(c.VkPipelineCacheCreateInfo);
    cache_info.sType = c.VK_STRUCTURE_TYPE_PIPELINE_CACHE_CREATE_INFO;

    // Try to load cache
    var file_buffer: []u8 = &[_]u8{};
    const file = std.fs.cwd().openFile("pipeline_cache.bin", .{}) catch null;
    if (file) |f| {
        defer f.close();
        if (f.stat()) |stat| {
            if (stat.size > 0) {
                const alloc = memory.cardinal_get_allocator_for_category(.RENDERER).as_allocator();
                if (alloc.alloc(u8, stat.size)) |buf| {
                    if (f.readAll(buf)) |read_bytes| {
                        if (read_bytes == stat.size) {
                            cache_info.initialDataSize = stat.size;
                            cache_info.pInitialData = buf.ptr;
                            file_buffer = buf;
                            pipe_log.info("Loading pipeline cache ({d} bytes)", .{stat.size});
                        }
                    } else |_| {}
                } else |_| {}
            }
        } else |_| {}
    }
    defer if (file_buffer.len > 0) {
        const alloc = memory.cardinal_get_allocator_for_category(.RENDERER).as_allocator();
        alloc.free(file_buffer);
    };

    if (c.vkCreatePipelineCache(s.context.device, &cache_info, null, &s.pipelines.pipeline_cache) != c.VK_SUCCESS) {
        pipe_log.err("Failed to create pipeline cache", .{});
        return false;
    }
    return true;
}

fn destroy_pipeline_cache(s: *types.VulkanState) void {
    if (s.pipelines.pipeline_cache != null) {
        // Save cache
        var size: usize = 0;
        if (c.vkGetPipelineCacheData(s.context.device, s.pipelines.pipeline_cache, &size, null) == c.VK_SUCCESS) {
            if (size > 0) {
                const alloc = memory.cardinal_get_allocator_for_category(.RENDERER).as_allocator();
                if (alloc.alloc(u8, size)) |buf| {
                    defer alloc.free(buf);
                    if (c.vkGetPipelineCacheData(s.context.device, s.pipelines.pipeline_cache, &size, buf.ptr) == c.VK_SUCCESS) {
                        if (std.fs.cwd().createFile("pipeline_cache.bin", .{})) |file| {
                            defer file.close();
                            _ = file.writeAll(buf) catch {};
                            pipe_log.info("Saved pipeline cache ({d} bytes)", .{size});
                        } else |_| {
                            pipe_log.warn("Failed to create cache file", .{});
                        }
                    }
                } else |_| {}
            }
        }

        c.vkDestroyPipelineCache(s.context.device, s.pipelines.pipeline_cache, null);
        s.pipelines.pipeline_cache = null;
    }
}

pub export fn vk_create_pipeline(s: ?*types.VulkanState) callconv(.c) bool {
    if (s == null) return false;
    const vs = s.?;

    // Create Pipeline Cache
    if (!create_pipeline_cache(vs)) {
        return false;
    }

    pipe_log.info("create depth resources", .{});
    if (!create_depth_resources(vs)) {
        return false;
    }

    pipe_log.info("pipeline: depth resources created - no simple triangle pipeline needed", .{});
    return true;
}

pub export fn vk_destroy_pipeline(s: ?*types.VulkanState) callconv(.c) void {
    if (s == null or s.?.context.device == null) return;
    const vs = s.?;

    // Wait for device to be idle before destroying resources for thread safety
    const result = c.vkDeviceWaitIdle(vs.context.device);
    if (result != c.VK_SUCCESS) {
        pipe_log.err("pipeline: vkDeviceWaitIdle failed during destruction: {d}", .{result});
        // Continue with destruction anyway to prevent resource leaks
    }

    destroy_pipeline_cache(vs);
    destroy_depth_resources(vs);

    pipe_log.info("pipeline resources destroyed", .{});
}
